sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestCollectorFactory.java

```diff
@@ -0,0 +1,290 @@
+        assertThat(rs, equalTo(Maybe.definitely("Coda")));
+
+        rs = dao.smartFindNameById(1);
+        assertThat(rs, equalTo(Maybe.definitely("Coda")));
+
+        rs = dao.inheritedGenericFindNameById(1);
+        assertThat(rs, equalTo(Maybe.definitely("Coda")));
+    }
+
+    @Test
+    public void testWithSqlObjectSetReturnValue() throws Exception {
+        Dao dao = SqlObjectBuilder.onDemand(dbi, Dao.class);
+        dao.insert(new Something(1, "Coda"));
+        dao.insert(new Something(2, "Brian"));
+
+        SortedSet<String> rs = dao.findAllAsSet();
+        assertThat(rs, equalTo(ImmutableSortedSet.of("Brian", "Coda")));
+    }
+
+
+    @RegisterContainerMapper({ImmutableListCollectorFactory.class, MaybeCollectorFactory.class})
+    public interface Dao extends Base<String> {
+        @SqlQuery("select name from something order by id")
+        ImmutableList<String> findAll();
+
+        @SqlQuery("select name from something order by id")
+        SortedSet<String> findAllAsSet();
+
+        @SqlUpdate("insert into something (id, name) values (:id, :name)")
+        void insert(@BindBean Something it);
+
+        @SqlQuery("select name from something where id = :id")
+        @SingleValueResult(String.class)
+        Maybe<String> findNameById(@Bind("id") int id);
+
+        @SqlQuery("select name from something where id = :id")
+        @SingleValueResult
+        Maybe<String> smartFindNameById(@Bind("id") int id);
+    }
+
+    public interface Base<T> {
+        @SqlQuery("select name from something where id = :id")
+        @SingleValueResult
+        Maybe<T> inheritedGenericFindNameById(@Bind("id") int id);
+    }
+
+    public static class ImmutableListCollectorFactory<T> implements CollectorFactory<T, ImmutableList.Builder<T>, ImmutableList<T>> {
+
+        public boolean accepts(Class<?> type) {
+            return ImmutableList.class.equals(type);
+        }
+
+        @Override
+        public Collector<T, ImmutableList.Builder<T>, ImmutableList<T>> newCollector(Class<ImmutableList<T>> type) {
+            return Collector.of(ImmutableList.Builder::new, ImmutableList.Builder::add, (first, second) -> {
+                throw new UnsupportedOperationException("Parallel collecting is not supported");
+            }, ImmutableList.Builder<T>::build);
+        }
+    }
+
+    public static class MaybeCollectorFactory<T> implements CollectorFactory<T, MaybeBuilder<T>, Maybe<T>> {
+
+        public boolean accepts(Class<?> type) {
+            return type.equals(Maybe.class);
+        }
+
+        @Override
+        public Collector<T, MaybeBuilder<T>, Maybe<T>> newCollector(Class<Maybe<T>> type) {
+            return Collector.of(MaybeBuilder::new, MaybeBuilder::set, (first, second) -> {
+                throw new UnsupportedOperationException("Parallel collecting is not supported");
+            }, MaybeBuilder::build);
+        }
+    }
+
+    public static abstract class Maybe<T> {
```

Comment:
Any reason to declare our own `Maybe<T>` here rather than just reuse `Optional<T>`?